File: Hard/question_6.js
// Abstraction is a fundamental concept in object-oriented programming that involves hiding the complex
// implementation details and showing only the necessary features of an object. It simplifies the complexity
// of a system, making it more understandable and manageable. A practical use case of abstraction is building
// a user interface (UI) library or framework where users interact with a simplified interface to create
// complex UI components.

// Abstract base class representing a UI component
class UIComponent {
  constructor() {
    if (new.target === UIComponent) {
      throw new Error(
        "UIComponent is an abstract class and cannot be instantiated directly."
      );
    }
  }

  // Abstract method to render the component
  render() {
    throw new Error("Render method must be implemented by concrete classes.");
  }

  // Common method for all UI components
  show() {
    console.log("UI component is shown.");
  }
}

// Concrete class representing a Button component
class Button extends UIComponent {
  constructor(label) {
    super();
    this.label = label;
  }

  // Implement the render method for the Button component
  render() {
    console.log(`Button with label: ${this.label} is rendered.`);
  }
}

// Concrete class representing a Textbox component
class Textbox extends UIComponent {
  constructor(value) {
    super();
    this.value = value;
  }

  // Implement the render method for the Textbox component
  render() {
    console.log(`Textbox with value: ${this.value} is rendered.`);
  }
}

// Usage of the UI library
const button = new Button("Submit");
button.render(); // Output: "Button with label: Submit is rendered."
button.show(); // Output: "UI component is shown."

const textbox = new Textbox("Type here");
textbox.render(); // Output: "Textbox with value: Type here is rendered."
textbox.show(); // Output: "UI component is shown."

File: Hard/question_6.py
# Abstraction explanation with an example:
# Abstraction is a fundamental concept in object-oriented programming that involves simplifying complex systems by modeling classes based on the essential properties and behaviors they share. It allows you to focus on the essential features of an object while hiding the unnecessary details.

# Practical Use Case of Abstraction: Database Interaction
# Consider a scenario where you are developing a web application that interacts with a database. Abstraction can be applied to create a generic database connection class that can be extended for specific database types (e.g., MySQL, PostgreSQL). Users of the application can then interact with the database without worrying about the underlying details.

from abc import ABC, abstractmethod

class DatabaseConnection(ABC):
    def __init__(self, connection_string):
        self._connection_string = connection_string

    @abstractmethod
    def connect(self):
        pass

    @abstractmethod
    def execute_query(self, query):
        pass

class MySQLConnection(DatabaseConnection):
    def connect(self):
        print(f"Connecting to MySQL database with {self._connection_string}")

    def execute_query(self, query):
        print(f"Executing MySQL query: {query}")

class PostgreSQLConnection(DatabaseConnection):
    def connect(self):
        print(f"Connecting to PostgreSQL database with {self._connection_string}")

    def execute_query(self, query):
        print(f"Executing PostgreSQL query: {query}")

# Example usage
mysql_db = MySQLConnection("mysql://user:password@localhost/mydatabase")
pgsql_db = PostgreSQLConnection("pgsql://user:password@localhost/mydatabase")

mysql_db.connect()
mysql_db.execute_query("SELECT * FROM users")

pgsql_db.connect()
pgsql_db.execute_query("SELECT * FROM employees")
